src/libos/src/net/socket/unix_socket/stream/socket.rs

```
@@ -0,0 +1,190 @@
+    pub fn new(flags: FileFlags) -> Result<Self> {
+        Ok(Self {
+            path: RwLock::new(UnixAddr::UnNamed),
+            end: SgxMutex::new(None),
+            is_blocking: AtomicBool::new(!flags.contains(FileFlags::SOCK_NONBLOCK)),
+        })
+    }
+
+    pub fn path(&self) -> UnixAddr {
+        self.path.read().unwrap().clone()
+    }
+
+    pub fn end(&self) -> &SgxMutex<Option<Arc<UnixEnd>>> {
+        &self.end
+    }
+
+    pub fn peer_name(&self) -> Result<UnixAddr> {
+        if let Some(end) = &*self.end.lock().unwrap() {
+            Ok(end.peer_name())
+        } else {
+            return_errno!(ENOTCONN, "the socket is not connected");
+        }
+    }
+
+    pub fn is_connected(&self) -> bool {
+        self.end.lock().unwrap().is_some()
+    }
+
+    pub fn is_blocking(&self) -> bool {
+        self.is_blocking.load(Ordering::SeqCst)
+    }
+
+    pub fn set_non_blocking(&self) {
+        self.is_blocking.store(false, Ordering::SeqCst);
+        let end = self.end.lock().unwrap();
+        end.as_ref().map(|c| c.set_non_blocking());
+    }
+
+    pub fn set_blocking(&self) {
+        self.is_blocking.store(true, Ordering::SeqCst);
+        let end = self.end.lock().unwrap();
+        end.as_ref().map(|c| c.set_blocking());
+    }
+
+    // TODO: create the corresponding file in the fs
+    pub fn bind(&self, addr: &UnixAddr) -> Result<()> {
+        if !self.path().is_unnamed() {
+            return_errno!(EINVAL, "the socket is already bound");
+        }
+
+        STREAM_SPACE.add_binder(addr)?;
+
+        *self.path.write().unwrap() = addr.clone();
+        if let Some(ref end) = *self.end.lock().unwrap() {
+            end.set_name(&addr);
+        }
+        Ok(())
+    }
+
+    //TODO: add backlog support
+    pub fn listen(&self, backlog: i32) -> Result<()> {
+        if self.path().is_unnamed() {
+            return_errno!(EINVAL, "the socket is not bound");
+        }
+
+        STREAM_SPACE.add_listener(&self.path())
```

Why if a socket listens _twice_? Should we throw an error?

---

Basically, this is for changing the backlog. We do not implement it but it is good to support the way this API is called.